src/Pages/Register.cshtml.cs

```diff
@@ -1,16 +1,62 @@
-using System;
 using System.Collections.Generic;
-using System.Linq;
+using System.ComponentModel.DataAnnotations;
+using System.Security.Claims;
 using System.Threading.Tasks;
+using ePiggyWeb.Authentication;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Authentication.Cookies;
 using Microsoft.AspNetCore.Mvc;
 using Microsoft.AspNetCore.Mvc.RazorPages;
 
 namespace ePiggyWeb.Pages
 {
     public class RegisterModel : PageModel
     {
-        public void OnGet()
+        [Required, EmailAddressAttribute(ErrorMessage = "Incorrect e-mail")]
+        [BindProperty]
+        public string Email { get; set; }
+
+        [Required]
+        [RegularExpression(@"^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@$!%*?&])[A-Za-z\d@$!%*?&]{6,}$", ErrorMessage = "Password must contain at least one uppercase letter, at least one number, special character and be longer than six characters.")]
+        [BindProperty]
+        [DataType(DataType.Password)]
+        public string Password { get; set; }
+
+        [Required]
+        [BindProperty]
+        [DataType(DataType.Password)]
+        public string PasswordConfirm { get; set; }
+
+        public string ErrorMessage = "";
+
+
+        public async Task<IActionResult> OnPost()
         {
+            if (!Password.Equals(PasswordConfirm))
+            {
+                ErrorMessage = "Passwords did not match!";
+                return Page();
+            }
+
+            var Id = UserAuth.Registration(Email, Password);
+            if (Id > -1)
+            {
+                 var claims = new List<Claim>
+                 {
+                     new Claim(ClaimTypes.UserData, Id.ToString()),
+                     new Claim(ClaimTypes.Email, Email)
+                 };
+                 var claimsIdentity = new ClaimsIdentity(claims, "Login");
+                 await HttpContext.SignInAsync(CookieAuthenticationDefaults.AuthenticationScheme,
+                     new ClaimsPrincipal(claimsIdentity));
+ 
+                 return Redirect("/Index");
+            };
```

Comment:
Could you explain how the registration works?  The things that caught my eye were 'async Task' and 'Claim',

Reply:
Citing: "An async method runs synchronously until it reaches its first await expression, at which point the method is suspended until the awaited task is complete. "
Claims are data that is saved in cookie and can be accessed as long as person logged in.
` var claims = new List<Claim>
                 {
                     new Claim(ClaimTypes.UserData, Id.ToString()),
                     new Claim(ClaimTypes.Email, Email)
                 };`
Creating list of claims (data saved in cookie.
`var claimsIdentity = new ClaimsIdentity(claims, "Login");`
Variable with specified claims, authentication type.
` await HttpContext.SignInAsync(CookieAuthenticationDefaults.AuthenticationScheme,
                     new ClaimsPrincipal(claimsIdentity));`
Creation of cookie (boilerplate code not sure about each step of it)
And finally: "IActionResult type. The IActionResult return type is appropriate when multiple ActionResult return types are possible in an action. The ActionResult types represent various HTTP status codes. Any non-abstract class deriving from ActionResult qualifies as a valid return type."

_Quite a bit to read, but I tried to be as informative as I could. 🙂 _

Reply:
I skipped some parts since I think they are understandable.

Reply:
Thanks for the in depth answer, clarified this method alot!